packages/internal/src/lib-nodejs/ava-unhandled-rejection.js

```diff
@@ -14,6 +14,104 @@ export const AVA_EXPECT_UNHANDLED_REJECTIONS =
 
 export const SUBTEST_PREFIX = '(unhandled rejection subprocess): ';
 
+const delayTurn = () => new Promise(resolve => setImmediate(resolve));
+const settleUnhandled = async () => {
+  await delayTurn();
+  await delayTurn();
+};
+
+const stripListeners = event => {
+  const listeners = process.listeners(event);
+  for (const listener of listeners) {
+    process.off(event, listener);
+  }
+  return () => {
+    for (const listener of listeners) {
+      process.on(event, listener);
+    }
+  };
+};
+
+const makeUnhandledTracker = () => {
+  let unhandled = 0;
+  let seenUnhandled = 0;
+  const pending = new Set();
+  const tokenByPromise = new WeakMap();
+  // Finalization callbacks run on the main thread; `pending` is only mutated
+  // here and after the flushes in countUnhandled, so no extra synchronization
+  // is required.
+  const registry = new FinalizationRegistry(token => {
+    if (pending.delete(token)) {
+      // We count once per promise when it is collected, regardless of whether
+      // a handler was attached later.
+      unhandled += 1;
+    }
+  });
+
+  const onUnhandled = (_reason, promise) => {
+    seenUnhandled += 1;
+    const token = Symbol('unhandledRejection');
+    tokenByPromise.set(promise, token);
+    pending.add(token);
+    registry.register(promise, token, token);
+  };
+
+  const onHandledLate = promise => {
+    tokenByPromise.delete(promise);
+    // Keep the token in `pending` so late handlers don't mask an earlier
+    // unhandled notification; we want to count every unhandledRejection event
+    // that fired in the subprocess even if it is observed later. The registry
+    // callback runs at most once per promise on the main thread, so this does
+    // not double-count. Tokens stay in `pending` only until the promise is GC'd
+    // and the finalizer fires to delete them.
+  };
+
+  const snapshot = () => ({ unhandled, pending: pending.size, seenUnhandled });
+
+  return {
+    onUnhandled,
+    onHandledLate,
+    snapshot,
+  };
+};
+
+/**
+ * Count unhandled rejections for the given work by running it after removing
+ * any existing listeners (like AVA's) and waiting for GC/finalization.
+ *
+ * @param {() => any | Promise<any>} work
+ * @param {object} powers
+ * @param {() => Promise<void>} powers.gcAndFinalize
+ */
+export const countUnhandled = async (work, { gcAndFinalize }) => {
```

Comment:
I suggest implementing the [`tameConsole` handling](https://github.com/endojs/endo/tree/master/packages/ses/src/error/tame-console.js) of this `unhandledRejectionTrapping` option.

```suggestion
 * @param {() => Promise<void>} powers.gcAndFinalize
 * @param {'report' | 'none'} [unhandledRejectionTrapping]
 */
export const countUnhandled = async (work, {
  gcAndFinalize,
  unhandledRejectionTrapping = 'report',
}) => {
```